### PR TITLE
ASL gesture recognition demo enhancement

### PR DESCRIPTION
**Pull request by evgeny-izutov**
- Added option to show example of some ASL gestures
- Fixed issue with multi-output network

### CODE REVIEW

**Comment by IRDonch** [comment]
I'm not quite sure what the value of this addition is. It appears to just show video files next to the main window. What does this demonstrate?

**Comment by evgeny-izutov** [comment]
@IRDonch, additional window demonstrates the example of ASL (American Sign Language) gesture which you can try to reproduce but not limited to show only this one gesture. It's made to support users who want to test the demo but don't know ASL. In future some video examples will be added too (after resolving some issues).

**Comment by IRDonch** [comment]
But all it does is play video files from a folder. If the user has this folder, they could just play the files themselves, no?

**Comment by evgeny-izutov** [comment]
Of course it plays video files from folder and nothing more... It's helper for the user. If you run this demo and try to show some gesture you will see nothing because ASL gesture is too different from random hand movement. As result such user will say that the demo doesn't work but it's not true. And as I said before video samples will be added in future and every user will have such folder with video samples.

**Comment by vladimir-dudnik** [comment]
@evgeny-izutov for me it look like should be enough to add pictures to model description with one may be two gestures which are recognized. That way you show customer how gesture should look like and that model cn actually recognize them. You should not convert demo to ASL teacher.

**Comment by vladimir-dudnik** [comment]
@evgeny-izutov even better, if you can add picture as GIF format with small demonstrating moview it would be perfect

**Comment by evgeny-izutov** [comment]
Ok, I open my cards and say that the demo is initially planned to have mode like "try to repeat" but to speed up the merging procedure of previous PR it was moved to this one PR. So I don't know what is the reason to discuss the planned early functionality. Also I can say that this video sample not for plain  users...

**Comment by vladimir-dudnik** [comment]
@evgeny-izutov you first of all should not be closing your cards. The reason to discuss 'planned early' functionality is that we may have a few days before release and that is not right time to introduce something new.

**Comment by evgeny-izutov** [comment]
Anyway I cannot understand why so many words for so simple but useful for live demo functionality have been said...This PR is about how to improve initial demo for the live scenario and nothing more...

**Comment by evgeny-izutov** [comment]
@snosov1

**Comment by snosov1** [comment]
I'm with Evgeny on this one - what's so controversial about the change? With the proposed change the demo is usable, say, at a conference booth (we play videos with correct gestures and ask the user to repeat them). Images (or GIFs) in documentation doesn't solve it, playing videos from the folder lacks interactivity. 

What do we risk with if we merge it?

**Comment by IRDonch** [comment]
> What do we risk with if we merge it?

This is the kind of thinking that feature creep is born from. :-) The risk, as with any other addition, is in the slowing down of maintenance. This addition might be fairly simple, sure, but simple additions add up.

The question that should be asked is what we _gain_ if we merge it. I believe the primary purpose of the demos is to demonstrate the usage of our models. Anything beyond that should be carefully considered. I am not wholly opposed to convenience features (the whole GUI is a convenience, really), but there is a line between user-friendliness and bloat. And since this addition doesn't do anything that couldn't be done by a video player with a playlist, I am unconvinced it does not belong to the latter category.

**Comment by snosov1** [comment]
> that couldn't be done by a video player with a playlist

"Video player with a playlist" sounds easier than it is on practice (unfortunately). Every platform has its own player, with its own peculiarities and writing a universal guide is more complicated than it sounds. (Say, what do we do on an APL with Yocto that we didn't setup ourselves?)

> The question that should be asked is what we gain if we merge it

We gain interactivity and easier maintenance on the "deployment" front (i.e. when we or other people actually show the demo live).

Anyway, I don't think we disagree on the basic principles. And it all just boils down to a judgement call - whether this feature gives us enough to validate itself or not. Our call (as the authors of the demo) is we should have it. Please, put any potential code maintenance bloat that it creates on us. 

If there's any other issues than that - we can discuss further. If not, please, accept.

File: demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        """Returns the path to the current video source"""
+
+        return self.source_paths[self.cur_source_id]
+
+    @staticmethod
+    def parse_source_paths(input_dir, valid_names):
+        """Returns the list of valid video sources"""
+
+        valid_names = [n.lower() for n in valid_names]
+        all_file_paths = [f for f in listdir(input_dir) if isfile(join(input_dir, f))]
```
[the remark this answers is not in the document]
Comment (evgeny-izutov) [reply demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py]: Fixed.

File: demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    def cur_source(self):
+        """Returns the path to the current video source"""
+
+        return self.source_paths[self.cur_source_id]
+
+    @staticmethod
+    def parse_source_paths(input_dir, valid_names):
+        """Returns the list of valid video sources"""
+
+        valid_names = [n.lower() for n in valid_names]
```
[the remark this answers is not in the document]
Comment (evgeny-izutov) [reply demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py]: Done.

File: demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        if self.cap is None:
+            self.cap = cv2.VideoCapture(source_path)
+
+        _, frame = self.cap.read()
+        if frame is None:
+            self.cap.set(cv2.CAP_PROP_POS_FRAMES, 0)
+            _, frame = self.cap.read()
+            assert frame is not None
+
+        if frame is not None:
```
[the remark this answers is not in the document]
Comment (evgeny-izutov) [reply demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py]: Fixed.

File: demos/python_demos/asl_recognition_demo/asl_recognition_demo.py [hunk 4]
```diff
@@ -109,6 +114,10 @@ def main():
     class_map = load_class_map(args.class_map)
     assert class_map is not None
 
+    samples_library = None
+    if args.samples_dir is not None and exists(args.samples_dir):
+      samples_library = VideoLibrary(args.samples_dir, SAMPLES_WINDOW_SIZE, list(class_map.values()))
```
[the remark this answers is not in the document]
Comment (evgeny-izutov) [reply demos/python_demos/asl_recognition_demo/asl_recognition_demo.py]: Fixed.

**Comment by IRDonch** [comment]
> Fixed issue with multi-output network

What multi-output network?

**Comment by evgeny-izutov** [comment]
> > Fixed issue with multi-output network
> 
> What multi-output network?

It's about person-detection-asl model. Each time when IR version is changed the name of output blob is changed too (question to MO). The most easiest way is to filter outputs by the shape of blob.

**Comment by IRDonch** [comment]
I see. In that case, could you submit this change as a separate PR? It looks like something that needs to be included in the release (whereas the rest of this PR looks like something that doesn't).

File: demos/python_demos/asl_recognition_demo/asl_recognition_demo/common.py [hunk 5]
```diff
@@ -45,19 +45,33 @@ def __init__(self, model_path, device, ie_core, num_requests, output_name=None):
...
-                self.output_name = output_name
+        if len(self.net.outputs) > 1:
+            if output_shape is not None:
+                candidates = []
+                for candidate_name in self.net.outputs:
+                    candidate_shape = self.exec_net.requests[0].outputs[candidate_name].shape
+                    if len(candidate_shape) != len(output_shape):
+                        continue
+
+                    matches = [src == trg if trg is not None and trg > 0 else True
```
Comment (IRDonch) [inline demos/python_demos/asl_recognition_demo/asl_recognition_demo/common.py]:
```suggestion
                    matches = [src == trg or trg < 0
```

It doesn't look like `trg` can be `None`.

Comment (evgeny-izutov) [reply demos/python_demos/asl_recognition_demo/asl_recognition_demo/common.py]: Fixed.

**Comment by evgeny-izutov** [comment]
> submit this change as a separate PR

For us both changes are equally significant and we expect to merge them before release. So I don't see the need in separate PR.

**Comment by IRDonch** [comment]
> For us both changes are equally significant and we expect to merge them before release.

How could they be equally significant? One is a new feature (and a convenience feature at that) and the other is an important fix.

I'm not merging new features when the release is just around the corner. Please be reasonable.

**Comment by evgeny-izutov** [comment]
> I'm not merging new features when the release is just around the corner

10 days we are discussing who is owner of the demo and who has the right to add changes in demo instead of merging and testing significantly early the release! And from our side (live demo opportunity) all this changes are critical.

**Comment by IRDonch** [comment]
The absolute majority of our users will not be showing this demo live. And if _you_ are going to show it live, then you could quite easily use a custom version of it. I really don't understand why you consider it necessary for this change to be included in the release.

**Comment by snosov1** [comment]
> I'm not merging new features when the release is just around the corner. Please be reasonable.

Roman, the whole demo is a new feature. None of the python demos have feature tests (only simple tests that they run) i.e. all the samples are equally broken except for the basic scenario. We're 10 days into this discussion. How "reasonable" is defined in your book? Cause I seem to have a very different meaning in mine. 

> The absolute majority of our users will not be showing this demo live.

Roman, if we're delaying the merge till after release, then, please:

1. Write the script that runs the app in "teacher" mode on Windows, Ubuntu and Yocto (i.e. downloads video player, runs videos in the loop, etc.)
2. Document the process and add it to readme
3. Implement tests for that functionality
4. Make sure that it passes a "hallway test" (we want to be a part of it).
5. Confirm that I can forward you all the colleagues/users that have issues with that process

**Comment by IRDonch** [comment]
Sergey,

First, there is no need for sarcasm. Please keep it civil.

Second, we are not 10 days into this discussion. Previously we have discussed the need for this feature at all, for which you have presented good arguments and convinced me. Currently, we are discussing the need for this feature to be in this release, for which I have not heard any arguments yet. The scenario that you described (showing the demo at a conference) does not require this feature to be in the release; you clearly need data that isn't in the OpenVINO toolkit to make use of it, so you could include your custom version of the demo in that data.

To be frank, given that this PR was opened when we had already started release stabilization and that it's just a convenience feature, I had always assumed that it was fine to delay until the next release. It is surprising to me that you're pushing for this so hard.

**Comment by snosov1** [comment]
> Currently, we are discussing the need for this feature to be in this release, for which I have not heard any arguments yet.

The main argument is that this feature is not "new". The whole demo is new. And without this feature it's only half-useful.

Likewise, I haven't heard arguments against including it, except for:

1. It's a video player (that you claim we've closed)
2. It puts a maintenance burden on you (that we're willing to accept ourselves)
3. The PR arrived after you've started "release stabilization". Pardon my ignorance, but I just don't know what that means. As far as I can see - there's no additional effort required from you or your team whether this change is in or out. Maybe I'm awfully wrong about that.

> so you could include your custom version of the demo in that data.

Sharing video files with 3rd parties has proven to be much easier than code/binaries. 

> It is surprising to me that you're pushing for this so hard.

Needless to say, we share the same astonishment by the resenting push.

**Comment by IRDonch** [comment]
> The main argument is that this feature is not "new". The whole demo is new. And without this feature it's only half-useful.

It's still new, though. It's not there now, and this PR adds it. That makes it new.

> The PR arrived after you've started "release stabilization". Pardon my ignorance, but I just don't know what that means.

It's a period during which we try to avoid merging new features, because there likely won't be enough time before the release to find and fix bugs in them.

> Sharing video files with 3rd parties has proven to be much easier than code/binaries.

Okay. I'll take that as justification.

**Comment by IRDonch** [comment]
@Wovchena Review, please.

**Comment by snosov1** [comment]
@IRDonch, thanks!

**Pull request merged**